cabal-install/Distribution/Client/InstallPlan.hs

```diff
@@ -641,3 +445,305 @@ configureInstallPlan solverPlan =
+-- installed already.
+--
+-- The result is both the packages that are now ready to be installed and also
+-- a 'Processing' state containing those same packages. The assumption is that
+-- all the packages that are ready will now be processed and so we can consider
+-- them to be in the processing state.
+--
+ready :: (HasUnitId ipkg, HasUnitId srcpkg)
+      => GenericInstallPlan ipkg srcpkg
+      -> ([GenericReadyPackage srcpkg], Processing)
+ready plan =
+    assert (processingInvariant plan processing) $
+    (readyPackages, processing)
+  where
+    !processing =
+      Processing
+        (IntSet.fromList
+          [ planVertexOf plan (installedUnitId pkg)
+          | pkg <- readyPackages ])
+        (IntSet.fromList
+          [ planVertexOf plan (installedUnitId pkg)
+          | PreExisting pkg <- toList plan ])
+        IntSet.empty
+    readyPackages =
+      [ ReadyPackage pkg
+      | Configured pkg <- toList plan
+      , hasAllInstalledDeps pkg
+      ]
+
+    hasAllInstalledDeps pkg =
+      all isPreExisting [ planPkgOf plan depv
+                        | let pkgv = planVertexOf plan (installedUnitId pkg)
+                        , depv <- planGraph plan ! pkgv ]
+    isPreExisting (PreExisting {}) = True
+    isPreExisting _                = False
+
+-- | Given a package in the processing state, mark the package as completed
+-- and return any packages that are newly in the processing state (ie ready to
+-- process), along with the updated 'Processing' state.
+--
+completed :: GenericInstallPlan ipkg srcpkg
+          -> Processing -> UnitId
+          -> ([GenericReadyPackage srcpkg], Processing)
+completed plan (Processing processingSet completedSet failedSet) pkgid =
+    assert (pkgv `IntSet.member` processingSet) $
+    assert (processingInvariant plan processing') $
+
+    ( map (asReadyPackage . planPkgOf plan) newlyReady
+    , processing' )
+  where
+    pkgv           = planVertexOf plan pkgid
+    completedSet'  = IntSet.insert pkgv completedSet
+
+    -- each direct reverse dep where all direct deps are completed
+    newlyReady     = [ depv
+                     | depv <- planGraphRev plan ! pkgv
+                     , all (`IntSet.member` completedSet')
+                           (planGraph plan ! depv)
+                     ]
+
+    processingSet' = foldl' (flip IntSet.insert)
+                            (IntSet.delete pkgv processingSet)
+                            newlyReady
+    processing'    = Processing processingSet' completedSet' failedSet
+
+    asReadyPackage (Configured pkg) = ReadyPackage pkg
+    asReadyPackage _ = error "InstallPlan.completed: internal error"
+
+failed :: GenericInstallPlan ipkg srcpkg
+       -> Processing -> UnitId
+       -> ([srcpkg], Processing)
+failed plan (Processing processingSet completedSet failedSet) pkgid =
+    assert (pkgv `IntSet.member` processingSet) $
+    assert (all (`IntSet.notMember` processingSet) (tail newlyFailed)) $
```

Comment:
I hate this, you're assuming that the first element returned by closure is the initial one, no reason to expect this to be the case from the docs. @dcoutts